In [4]:
%%time
import pandas as pd

pandass = pd.read_csv('large.csv')
pandass.head()

<string>:2: DtypeWarning: Columns (17,18,20,21,22,23,29,30,31,32,34,36,38,39) have mixed types.Specify dtype option on import or set low_memory=False.


Wall time: 52.8 s


Summons Number Plate ID Registration State Plate Type  Issue Date  \
0      1363745270  GGY6450                 99        PAS  07/09/2015   
1      1363745293   KXD355                 SC        PAS  07/09/2015   
2      1363745438  JCK7576                 PA        PAS  07/09/2015   
3      1363745475  GYK7658                 NY        OMS  07/09/2015   
4      1363745487  GMT8141                 NY        PAS  07/09/2015   

   Violation Code Vehicle Body Type Vehicle Make Issuing Agency  Street Code1  \
0              46               SDN        HONDA              P             0   
1              21              SUBN        CHEVR              P         55730   
2              21               SDN        ME/BE              P         42730   
3              21              SUBN        NISSA              P         58130   
4              21               P-U        LINCO              P         58130   

   ...  Hydrant Violation  Double Parking Violation  Latitude  Longitude  \
0  ...                NaN                       NaN       NaN        NaN   
1  ...                NaN                       NaN       NaN        NaN   
2  ...                NaN                       NaN       NaN        NaN   
3  ...                NaN                       NaN       NaN        NaN   
4  ...                NaN                       NaN       NaN        NaN   

   Community Board  Community Council   Census Tract BIN BBL NTA  
0              NaN                 NaN           NaN NaN NaN NaN  
1              NaN                 NaN           NaN NaN NaN NaN  
2              NaN                 NaN           NaN NaN NaN NaN  
3              NaN                 NaN           NaN NaN NaN NaN  
4              NaN                 NaN           NaN NaN NaN NaN  

[5 rows x 51 columns]

In [3]:
%%time
import dask.dataframe as dd

ddask = dd.read_csv('large.csv')
ddask

Wall time: 27 ms


Dask DataFrame Structure:
               Summons Number Plate ID Registration State Plate Type Issue Date Violation Code Vehicle Body Type Vehicle Make Issuing Agency Street Code1 Street Code2 Street Code3 Vehicle Expiration Date Violation Location Violation Precinct Issuer Precinct Issuer Code Issuer Command Issuer Squad Violation Time Time First Observed Violation County Violation In Front Of Or Opposite House Number Street Name Intersecting Street Date First Observed Law Section Sub Division Violation Legal Code Days Parking In Effect     From Hours In Effect To Hours In Effect Vehicle Color Unregistered Vehicle? Vehicle Year Meter Number Feet From Curb Violation Post Code Violation Description No Standing or Stopping Violation Hydrant Violation Double Parking Violation Latitude Longitude Community Board Community Council  Census Tract      BIN      BBL      NTA
npartitions=34                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     
                        int64   object             object     object     object          int64            object       object         object        int64        int64        int64                   int64            float64              int64           int64       int64         object        int64         object              object           object                            object       object      object              object               int64       int64       object              float64                     object               object             object        object               float64        int64       object          int64             float64               float64                           float64           float64                  float64  float64   float64         float64            float64      float64  float64  float64  float64
                          ...      ...                ...        ...        ...            ...               ...          ...            ...          ...          ...          ...                     ...                ...                ...             ...         ...            ...          ...            ...                 ...              ...                               ...          ...         ...                 ...                 ...         ...          ...                  ...                        ...                  ...                ...           ...                   ...          ...          ...            ...                 ...                   ...                               ...               ...                      ...      ...       ...             ...                ...          ...      ...      ...      ...
...                       ...      ...                ...        ...        ...            ...               ...          ...            ...          ...          ...          ...                     ...                ...                ...             ...         ...            ...          ...            ...                 ...              ...                               ...          ...         ...                 ...                 ...         ...          ...                  ...                        ...                  ...                ...     

In [5]:
%%time
import modin.pandas as md
import ray 
# ray.init()
modinn = md.read_csv('large.csv')
modinn

To remove this warning, run the following python code before doing dataframe operations:

    import ray
    ray.init()



Wall time: 41.1 s


Summons Number Plate ID Registration State Plate Type  Issue Date  \
0             1363745270  GGY6450                 99        PAS  07/09/2015   
1             1363745293   KXD355                 SC        PAS  07/09/2015   
2             1363745438  JCK7576                 PA        PAS  07/09/2015   
3             1363745475  GYK7658                 NY        OMS  07/09/2015   
4             1363745487  GMT8141                 NY        PAS  07/09/2015   
...                  ...      ...                ...        ...         ...   
10626894      8359500927  FFM1969                 NY        PAS  06/08/2016   
10626895      8359500940  2482082                 IN        PAS  06/08/2016   
10626896      8359500952  65919JW                 NY        COM  06/08/2016   
10626897      8359500964  68718MG                 NY        COM  06/08/2016   
10626898      8359500976  FZX4974                 NY        PAS  06/08/2016   

          Violation Code Vehicle Body Type Vehicle Make Issuing Agency  \
0                     46               SDN        HONDA              P   
1                     21              SUBN        CHEVR              P   
2                     21               SDN        ME/BE              P   
3                     21              SUBN        NISSA              P   
4                     21               P-U        LINCO              P   
...                  ...               ...          ...            ...   
10626894              38              SUBN        TOYOT              T   
10626895              51               VAN        FRUEH              T   
10626896              52               VAN        FRUEH              T   
10626897              82              PICK        DODGE              T   
10626898              38              4DSD        HONDA              T   

          Street Code1  ...  Hydrant Violation  Double Parking Violation  \
0                    0  ...                NaN                       NaN   
1                55730  ...                NaN                       NaN   
2                42730  ...                NaN                       NaN   
3                58130  ...                NaN                       NaN   
4                58130  ...                NaN                       NaN   
...                ...  ...                ...                       ...   
10626894         25590  ...                NaN                       NaN   
10626895         73980  ...                NaN                       NaN   
10626896             0  ...                NaN                       NaN   
10626897             0  ...                NaN                       NaN   
10626898             0  ...                NaN                       NaN   

          Latitude  Longitude  Community Board  Community Council   \
0              NaN        NaN              NaN                 NaN   
1              NaN        NaN              NaN                 NaN   
2              NaN        NaN              NaN                 NaN   
3              NaN        NaN              NaN                 NaN   
4              NaN        NaN              NaN                 NaN   
...            ...        ...              ...                 ...   
10626894       NaN        NaN              NaN                 NaN   
10626895       NaN        NaN              NaN                 NaN   
10626896       NaN        NaN              NaN                 NaN   
10626897       NaN        NaN              NaN                 NaN   
10626898       NaN        NaN              NaN                 NaN   

          Census Tract  BIN  BBL  NTA  
0                  NaN  NaN  NaN  NaN  
1                  NaN  NaN  NaN  NaN  
2                  NaN  NaN  NaN  NaN  
3                  NaN  NaN  NaN  NaN  
4                  NaN  NaN  NaN  NaN  
...                ...  ...  ...  ...  
10626894           NaN  NaN  NaN  NaN  
10626895           NaN  NaN  NaN  NaN  
10626896           NaN  NaN  NaN  NaN  
10626897          

In [6]:
ray.shutdown()

# In term of computer efficiency:
1. Dask 27 ms<br>
2. Modin with Ray 41.1 s <br>
3. Pandas 52.8 s <br>


In [7]:
ddask.columns

Index(['Summons Number', 'Plate ID', 'Registration State', 'Plate Type',
       'Issue Date', 'Violation Code', 'Vehicle Body Type', 'Vehicle Make',
       'Issuing Agency', 'Street Code1', 'Street Code2', 'Street Code3',
       'Vehicle Expiration Date', 'Violation Location', 'Violation Precinct',
       'Issuer Precinct', 'Issuer Code', 'Issuer Command', 'Issuer Squad',
       'Violation Time', 'Time First Observed', 'Violation County',
       'Violation In Front Of Or Opposite', 'House Number', 'Street Name',
       'Intersecting Street', 'Date First Observed', 'Law Section',
       'Sub Division', 'Violation Legal Code', 'Days Parking In Effect    ',
       'From Hours In Effect', 'To Hours In Effect', 'Vehicle Color',
       'Unregistered Vehicle?', 'Vehicle Year', 'Meter Number',
       'Feet From Curb', 'Violation Post Code', 'Violation Description',
       'No Standing or Stopping Violation', 'Hydrant Violation',
       'Double Parking Violation', 'Latitude', 'Longitude', 'Comm

In [8]:
ddask.columns=ddask.columns.str.strip().str.lower().str.replace(' ', '_')

In [9]:
ddask.columns

Index(['summons_number', 'plate_id', 'registration_state', 'plate_type',
       'issue_date', 'violation_code', 'vehicle_body_type', 'vehicle_make',
       'issuing_agency', 'street_code1', 'street_code2', 'street_code3',
       'vehicle_expiration_date', 'violation_location', 'violation_precinct',
       'issuer_precinct', 'issuer_code', 'issuer_command', 'issuer_squad',
       'violation_time', 'time_first_observed', 'violation_county',
       'violation_in_front_of_or_opposite', 'house_number', 'street_name',
       'intersecting_street', 'date_first_observed', 'law_section',
       'sub_division', 'violation_legal_code', 'days_parking_in_effect',
       'from_hours_in_effect', 'to_hours_in_effect', 'vehicle_color',
       'unregistered_vehicle?', 'vehicle_year', 'meter_number',
       'feet_from_curb', 'violation_post_code', 'violation_description',
       'no_standing_or_stopping_violation', 'hydrant_violation',
       'double_parking_violation', 'latitude', 'longitude', 'communit

In [33]:
import yaml

config = [
    {
        'read': {
                    'file_name': 'large.csv',
                    'delimiter': ',',
                    'columns': ['summons_number', 'plate_id', 'registration_state', 'plate_type',
                       'issue_date', 'violation_code', 'vehicle_body_type', 'vehicle_make',
                       'issuing_agency', 'street_code1', 'street_code2', 'street_code3',
                       'vehicle_expiration_date', 'violation_location', 'violation_precinct',
                       'issuer_precinct', 'issuer_code', 'issuer_command', 'issuer_squad',
                       'violation_time', 'time_first_observed', 'violation_county',
                       'violation_in_front_of_or_opposite', 'house_number', 'street_name',
                       'intersecting_street', 'date_first_observed', 'law_section',
                       'sub_division', 'violation_legal_code', 'days_parking_in_effect',
                       'from_hours_in_effect', 'to_hours_in_effect', 'vehicle_color',
                       'unregistered_vehicle?', 'vehicle_year', 'meter_number',
                       'feet_from_curb', 'violation_post_code', 'violation_description',
                       'no_standing_or_stopping_violation', 'hydrant_violation',
                       'double_parking_violation', 'latitude', 'longitude', 'community_board',
                       'community_council', 'census_tract', 'bin', 'bbl', 'nta']
                }, 
        
        'write': { 
            'file_name': 'large_pipe.gz',
            'delimiter': '|',
            'compression':'gzip',
            'columns': ['summons_number', 'plate_id', 'registration_state', 'plate_type',
       'issue_date', 'violation_code', 'vehicle_body_type', 'vehicle_make',
       'issuing_agency', 'street_code1', 'street_code2', 'street_code3',
       'vehicle_expiration_date', 'violation_location', 'violation_precinct',
       'issuer_precinct', 'issuer_code', 'issuer_command', 'issuer_squad',
       'violation_time', 'time_first_observed', 'violation_county',
       'violation_in_front_of_or_opposite', 'house_number', 'street_name',
       'intersecting_street', 'date_first_observed', 'law_section',
       'sub_division', 'violation_legal_code', 'days_parking_in_effect',
       'from_hours_in_effect', 'to_hours_in_effect', 'vehicle_color',
       'unregistered_vehicle?', 'vehicle_year', 'meter_number',
       'feet_from_curb', 'violation_post_code', 'violation_description',
       'no_standing_or_stopping_violation', 'hydrant_violation',
       'double_parking_violation', 'latitude', 'longitude', 'community_board',
       'community_council', 'census_tract', 'bin', 'bbl', 'nta']
                    }
    }
]

with open("config.yaml", 'w') as yamlfile:
    data = yaml.dump(config, yamlfile)
    print("Write successful")


Write successful


In [13]:
with open("config.yaml", "r") as yamlfile:
    config = yaml.load(yamlfile)
    print("Read successful")
print(config)

Read successful
[{'read': {'columns': ['summons_number', 'plate_id', 'registration_state', 'plate_type', 'issue_date', 'violation_code', 'vehicle_body_type', 'vehicle_make', 'issuing_agency', 'street_code1', 'street_code2', 'street_code3', 'vehicle_expiration_date', 'violation_location', 'violation_precinct', 'issuer_precinct', 'issuer_code', 'issuer_command', 'issuer_squad', 'violation_time', 'time_first_observed', 'violation_county', 'violation_in_front_of_or_opposite', 'house_number', 'street_name', 'intersecting_street', 'date_first_observed', 'law_section', 'sub_division', 'violation_legal_code', 'days_parking_in_effect', 'from_hours_in_effect', 'to_hours_in_effect', 'vehicle_color', 'unregistered_vehicle?', 'vehicle_year', 'meter_number', 'feet_from_curb', 'violation_post_code', 'violation_description', 'no_standing_or_stopping_violation', 'hydrant_violation', 'double_parking_violation', 'latitude', 'longitude', 'community_board', 'community_council', 'census_tract', 'bin', 'bbl'

# Read using config file

In [20]:
pandass = pd.read_csv(config[0]['read']['file_name'], delimiter = config[0]['read']['delimiter'], names = config[0]['read']['columns'])
pandass.head()

summons_number  plate_id  registration_state  plate_type  issue_date  \
0  Summons Number  Plate ID  Registration State  Plate Type  Issue Date   
1      1363745270   GGY6450                  99         PAS  07/09/2015   
2      1363745293    KXD355                  SC         PAS  07/09/2015   
3      1363745438   JCK7576                  PA         PAS  07/09/2015   
4      1363745475   GYK7658                  NY         OMS  07/09/2015   

   violation_code  vehicle_body_type  vehicle_make  issuing_agency  \
0  Violation Code  Vehicle Body Type  Vehicle Make  Issuing Agency   
1              46                SDN         HONDA               P   
2              21               SUBN         CHEVR               P   
3              21                SDN         ME/BE               P   
4              21               SUBN         NISSA               P   

   street_code1  ...  hydrant_violation  double_parking_violation  latitude  \
0  Street Code1  ...  Hydrant Violation  Double Parking Violation  Latitude   
1             0  ...                NaN                       NaN       NaN   
2         55730  ...                NaN                       NaN       NaN   
3         42730  ...                NaN                       NaN       NaN   
4         58130  ...                NaN                       NaN       NaN   

   longitude  community_board   community_council  census_tract  bin  bbl  nta  
0  Longitude  Community Board  Community Council   Census Tract  BIN  BBL  NTA  
1        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  
2        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  
3        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  
4        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  

[5 rows x 51 columns]

# Number of columns validation

In [22]:
len(pandass.columns)== len(config[0]['read']['columns'])

True

# Column name validation

In [23]:
all( x in config[0]['read']['columns'] for x in pandass.columns)

True

# Write with pipe delimiter

In [34]:
pandass.to_csv(config[0]['write']['file_name'], sep = config[0]['write']['delimiter'],
          compression=config[0]['write']['compression']) 

check

In [35]:
pipe_file = pd.read_csv('large_pipe.gz', delimiter= '|')
pipe_file.head()

Unnamed: 0  summons_number  plate_id  registration_state  plate_type  \
0           0  Summons Number  Plate ID  Registration State  Plate Type   
1           1      1363745270   GGY6450                  99         PAS   
2           2      1363745293    KXD355                  SC         PAS   
3           3      1363745438   JCK7576                  PA         PAS   
4           4      1363745475   GYK7658                  NY         OMS   

   issue_date  violation_code  vehicle_body_type  vehicle_make  \
0  Issue Date  Violation Code  Vehicle Body Type  Vehicle Make   
1  07/09/2015              46                SDN         HONDA   
2  07/09/2015              21               SUBN         CHEVR   
3  07/09/2015              21                SDN         ME/BE   
4  07/09/2015              21               SUBN         NISSA   

   issuing_agency  ...  hydrant_violation  double_parking_violation  latitude  \
0  Issuing Agency  ...  Hydrant Violation  Double Parking Violation  Latitude   
1               P  ...                NaN                       NaN       NaN   
2               P  ...                NaN                       NaN       NaN   
3               P  ...                NaN                       NaN       NaN   
4               P  ...                NaN                       NaN       NaN   

   longitude  community_board   community_council  census_tract  bin  bbl  nta  
0  Longitude  Community Board  Community Council   Census Tract  BIN  BBL  NTA  
1        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  
2        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  
3        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  
4        NaN              NaN                 NaN           NaN  NaN  NaN  NaN  

[5 rows x 52 columns]

# Summary

In [31]:
import os 

print('Number of rows:', pandass.shape[0])
print('Number of columns:', pandass.shape[1])
print('Orginal file size:', os.path.getsize('large.csv')/(10**9), 'GB')
print('Compressed file size:', os.path.getsize('large_pipe.csv.gz')/(10**9) , 'GB')

Number of rows: 10626900
Number of columns: 51
Orginal file size: 2.151937808 GB
Compressed file size: 0.428182073 GB
